backend/app/anti_recommendation_engine/anti_recommendation_engine.py

```diff
@@ -57,18 +76,26 @@ def generate_anti_recommendation_records(
                 self.__stack.append(self.__current_anti_recommendation_records)
 
             self.__current_anti_recommendation_records = [
-                self.__record_store[record_key],
+                self.__records_by_key[record_key],
                 *records_of_anti_recommendations,
             ]
 
         return tuple(records_of_anti_recommendations)
 
-    def get_previous_anti_recommendation_records(
+    def get_previous_records_of_anti_recommendations(
         self,
-    ) -> tuple[Record, ...] | None:
-        """Return a tuple of Records that have the same titles as previous AntiRecommendations"""
+    ) -> tuple[Record | None, ...]:
+        """Return a tuple of Records that were the previous Records of AntiRecommendations."""
 
         if self.__stack:
             return tuple(self.__stack.pop())
 
-        return None
+        return (None,)
+
+    def generate_anti_recommendations(
```

Comment:
If no one external consumer is calling this then you don't really need to expose it / subclass `AntiRecommender`.

Reply:
`AntiRecommendationEngine` is no longer subclassing `AntiRecommender`.